From 3e1fc6d3b44d5706b8e24f0c86226740c451f812 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 18:59:42 -0400
Subject: [fs] btrfs: Add treelog mount option

Message-id: <1407179109-13714-96-git-send-email-sandeen@redhat.com>
Patchwork-id: 87013
O-Subject: [RHEL7.1 PATCH 095/422] btrfs: Add treelog mount option.
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Qu Wenruo <quwenruo@cn.fujitsu.com>

commit a88998f291fc707f18ee42ae45220a3a3e384c27
Author: Qu Wenruo <quwenruo@cn.fujitsu.com>
Date:   Mon Jan 6 09:58:33 2014 +0800

    btrfs: Add treelog mount option.

    Add treelog mount option to enable tree log with
    remount option.

    Signed-off-by: Qu Wenruo <quwenruo@cn.fujitsu.com>
    Reviewed-by: David Sterba <dsterba@suse.cz>
    Signed-off-by: Josef Bacik <jbacik@fb.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/Documentation/filesystems/btrfs.txt b/Documentation/filesystems/btrfs.txt
index ab627e0..f97b4dc 100644
--- a/Documentation/filesystems/btrfs.txt
+++ b/Documentation/filesystems/btrfs.txt
@@ -164,8 +164,9 @@ Options with (*) are default options and will not show in the mount options.
  Enable/disable data checksumming for newly created files.
  Datasum implies datacow.
 
+  treelog(*)
   notreelog
- Disable the tree logging used for fsync and O_SYNC writes.
+ Enable/disable the tree logging used for fsync and O_SYNC writes.
 
   recovery
  Enable autorecovery attempts if a bad tree root is found at mount time.
diff --git a/fs/btrfs/super.c b/fs/btrfs/super.c
index e84e6cb..16d7fc7 100644
--- a/fs/btrfs/super.c
+++ b/fs/btrfs/super.c
@@ -325,7 +325,7 @@ enum {
  Opt_check_integrity_print_mask, Opt_fatal_errors, Opt_rescan_uuid_tree,
  Opt_commit_interval, Opt_barrier, Opt_nodefrag, Opt_nodiscard,
  Opt_noenospc_debug, Opt_noflushoncommit, Opt_acl, Opt_datacow,
- Opt_datasum,
+ Opt_datasum, Opt_treelog,
  Opt_err,
 };
 
@@ -353,6 +353,7 @@ static match_table_t tokens = {
  {Opt_acl, "acl"},
  {Opt_noacl, "noacl"},
  {Opt_notreelog, "notreelog"},
+ {Opt_treelog, "treelog"},
  {Opt_flushoncommit, "flushoncommit"},
  {Opt_noflushoncommit, "noflushoncommit"},
  {Opt_ratio, "metadata_ratio=%d"},
@@ -579,6 +580,11 @@ int btrfs_parse_options(struct btrfs_root *root, char *options)
    btrfs_info(root->fs_info, "disabling tree log");
    btrfs_set_opt(info->mount_opt, NOTREELOG);
    break;
+  case Opt_treelog:
+   if (btrfs_test_opt(root, NOTREELOG))
+    btrfs_info(root->fs_info, "enabling tree log");
+   btrfs_clear_opt(info->mount_opt, NOTREELOG);
+   break;
   case Opt_flushoncommit:
    btrfs_info(root->fs_info, "turning on flush-on-commit");
    btrfs_set_opt(info->mount_opt, FLUSHONCOMMIT);
-- 
1.7.1